src/main/java/com/zendesk/maxwell/producer/MaxwellNatsProducer.java

```diff
@@ -0,0 +1,229 @@
+	private final ArrayBlockingQueue<RowMap> queue;
+	private Thread thread;
+
+
+	public MaxwellNatsProducerWorker(MaxwellContext context, String natsSubject, String clusterId,
+					 String clientId, ArrayBlockingQueue<RowMap> queue) throws IOException, InterruptedException {
+		super(context);
+
+		this.natsSubject = natsSubject;
+		this.queue = queue;
+		this.connectionListener = new NatsConnectionListener(context);
+
+		String natsUrl = context.getConfig().natsUrl;
+		Long pubAckTimeout = context.getConfig().natsPubAckTimeout;
+
+		Options.Builder clientOptions = new Options.Builder();
+		io.nats.streaming.Options.Builder streamingOptions = new io.nats.streaming.Options.Builder();
+
+		clientOptions
+			.server(natsUrl)
+			.connectionListener(connectionListener);
+
+		if (context.getConfig().natsMaxReconnects > 0) {
+			clientOptions.maxReconnects(context.getConfig().natsMaxReconnects);
+		}
+
+		if (context.getConfig().natsReconnectWait != null) {
+			clientOptions.reconnectWait(Duration.ofMillis(context.getConfig().natsReconnectWait));
+		}
+
+		this.natsConnection = Nats.connect(clientOptions.build());
+		streamingOptions.natsConn(natsConnection);
+
+		if (pubAckTimeout != null) {
+			streamingOptions.pubAckWait(Duration.ofMillis(pubAckTimeout));
+		}
+
+		this.streamingConnection = NatsStreaming.connect(clusterId, clientId, streamingOptions.build());
+
+	}
+
+	@Override
+	public void run() {
+		this.thread = Thread.currentThread();
+		while ( true ) {
+			try {
+				//Irrespective of Nats is working or not, heartbeat should be pushed
+				if (queue.peek() instanceof HeartbeatRowMap) {
+					RowMap row = queue.take();
+					this.push(row);
+				} else if (connectionListener.isNatsRunning()){
+					RowMap row = queue.take();
+					this.push(row);
+				}
+
+			} catch ( Exception e ) {
+				LOGGER.info("Error: ", e);
+			}
+		}
+	}
+
+	@Override
+	public void sendAsync(RowMap r, AbstractAsyncProducer.CallbackCompleter cc) throws Exception {
+		String msg = r.toJSON(outputConfig);
+		NatsAckHandler handler = new NatsAckHandler(cc, r.getNextPosition(), msg,
+			this.succeededMessageCount, this.failedMessageCount, this.succeededMessageMeter, this.failedMessageMeter, this.context);
+		streamingConnection.publish(natsSubject, msg.getBytes(), handler);
+	}
+
+}
+
+class NatsConnectionListener implements ConnectionListener {
+	static final Logger LOGGER = LoggerFactory.getLogger(MaxwellNatsProducer.class);
+	private Boolean isNatsRunning = true;
+	private final MaxwellContext context;
+
+	public NatsConnectionListener(MaxwellContext context) {
+		this.context = context;
+	}
+
+	@Override
+	public void connectionEvent(Connection connection, Events events) {
+		switch (events) {
+			case DISCONNECTED:
+				isNatsRunning = false;
+				LOGGER.error("Nats connection disconnected");
+				if (!context.getConfig().ignoreProducerError) {
+					this.context.terminate(new RuntimeException("Nats Disconnected"));
+					return;
```

Comment:
should this be something that crashes maxwell?  Or can we wait for a reconnect?

Reply:
if ignoreProducerError is false, then we should terminate right?

Reply:
I'm just not sure what the `DISCONNECTED` message signifies: does it mean the client disconnected and will try to re-open the connection, or does it mean that it's dead and has given up?

Reply:
DISCONNECTED callback comes when the nats connection was down.